### ISSUE TITLE
Support TURN server without TLS cert

### ISSUE THREAD

**Issue by alvinthen** [livekit/livekit-server#71]
Hi, we're trying to understand why is the TLS cert required for the TURN server.

We're considering deploying in DigitalOcean with k8s and their load balancer. The load balancer will terminate the SSL before it reaches the cluster.

However, following the Helm chart and also the server source code, it seems like the TLS cert is necessary. Is there any way to run TURN without the cert?

**Comment by davidzhao** [comment]
We talk about this a bit in [deployment docs](https://docs.livekit.io/guides/deploy#turn). TURN is used primarily in three scenarios:

1. Both peers are behind symmetrical NAT (doesn't apply since LiveKit server acts as SFU)
2. Firewall blocks UDP (LiveKit uses ICE/TCP, and all major browsers/native SDKs support this)
3. Firewall blocks TCP traffic that's not TLS/443. (TURN/TLS is the only way for it to connect)

Case 3 is what we are addressing with the embedded TURN/TLS setup. Do you see any other cases where non-TLS TURN would be helpful?

We'll also support TURN/UDP as well, as corporate firewalls are increasingly open to UDP traffic over 443 due to QUIC. Server has support for this today, but we still need to update the helm chart to make it more easily deployable.

**Comment by alvinthen** [comment]
Absolutely understand the need for TLS. I'm trying to wrap my head around terminating TLS at the external load balancer, does it not work that way?

With my lack of understanding for k8s, I was assuming I could do

```
Clients -> External Load Balancer (DO Load Balancer, TLS terminated here) -> Kubernetes cluster -> Ingress -> LiveKit/TURN
```

Or must I use a cert-manager to supply the certs, as in

```
Clients -> Kubernetes cluster -> Ingress -> LiveKit/TURN (TLS terminated here)
```

**Comment by davidzhao** [comment]
Today, we only support the latter. Technically, terminating TLS at load balancer should work ok too. We picked one approach initially to avoid the complexity & confusion. Since not L4 load balancers support TLS termination, we went with the more compatible approach.

We can definitely consider doing the former too. But IMHO the performance gain is minimal: majority of your traffic would go straight through UDP.

Are you deploying to DO? We'd love help & contributions with making our helm chart DO compatible.

**Comment by alvinthen** [comment]
Thanks for clearing it up! And yes, we're deploying to DO, and with cert-manager so we can get certs from Let's Encrypt. It'll be great if you can help on the helm chart. We're still studying the helm chart and got confused on the load balancer part, hence this thread.

Thanks again!

**Comment by jfey** [comment]
Hello. We plan to check out livekit soon as part of a larger system. We are currently using Traefik as the load balancer, cert handler and TLS termination endpoint. Later we want to use k8s for the final rollouts. Thus we would also have the need to have TURN running behind Traefik without TLS.

Clients -(TLS) -> Traefik -> (non TLS) -> microservices and livekit, TURN etc.

I guess i have to watch this discussion here. Maybe we can run TURN outside of the box for a while with TLS enabled etc.

Thanks for all the clarifications

**Comment by hn8** [comment]
Livekit's TURN is expected to talk to SFU in the same host. If not on the same pod, may need to make sure UDP port 1024-30000 for TURN is accessible. Ideally clients should talk to TURN directly without going through ingress. Not sure TLS terminated in LB would work actually, but TCP proxy works. But if you plan to use the same LB for both WS and TURN, you may need to use 2 SNIs so LB can distinguish two traffic in the same port 443.

As David suggested, TURN-UDP (only port 53/80/443 is allow if <1024) and ICE-TCP (only port 80/443 is allowed if <1024) may be sufficient for most cases. For TURN-TLS, assigning different hostname with wildcard TLS cert is probaby easier than figure out how to make TURN-TLS work behind ingress/LB.

**Comment by davidzhao** [comment]
@jfey you can still use Traefik in front of LiveKit TURN, it's just that you would leave the termination to LiveKit (for now). If you are thinking about routing both TURN/TLS and HTTPS on the same load balancer/same port, I think it would be rather tricky to set up.

**Comment by jfey** [comment]
@ davidzhao: Understood. I am sure that this discussion already saved a lot of time on our side. We can re-arrange things and since this we are still in an early prepping phase there were no animals hurt till now. A lot to learn still. Wonderful community here.

**Comment by lukasIO** [comment]
@davidzhao I'd be happy to give it a try to implement a TCP listener, when using a loadbalancer that handles TLS termination.
I had a look at the pion/turn examples but would still need your input on how to best tackle it. Some scenarios I can think of:
- go for another config option like `tcpPort` that can be configured alternatively to `tlsPort` but `tlsPort` would take precedence if both are present.
- change `tlsPort` to a more general `tcpPort` that automatically uses tls when certificates are supplied and falls back to plain tcp if there are none.
- keep `tlsPort` and add a `allowInsecure` property which if set to `true` would look for the supplied certificates, but if none are present it would open a plain tcp listener instead of the tls one.

**Comment by hn8** [comment]
@lukasIO TCP ICE may more efficient thant TCP TURN in the case that UDP is blocked and TLS cert is not available.

**Comment by davidzhao** [comment]
I think with external termination, this makes a lot of sense. I think of all of the options you've proposed, I like a variant of 3: keep `tlsPort`, and add `externalTLS bool`. When `externalTLS` is set to true, it'll skip looking for certificates and listen to it as plain TCP. *but* still send back the candidate as a TURN/TLS candidate.

**Comment by hn8** [comment]
@davidzhao @lukasIO Is it possible to terminate TLS and forward TCP for TURN since it is not HTTP? I could not make it work with Envoy before. Does it need special kind of filter or like a TCP proxy with TLS termination (it means port need to be the same)?

**Comment by lukasIO** [comment]
> 
> 
> @davidzhao @lukasIO Is it possible to terminate TLS and forward TCP for TURN since it is not HTTP? I could not make it work with Envoy before, only TLS forwarding (with SNI) works. Does it need special kind of protocol filter? Cloud load balancer only supports HTTP for TLS termination AFAIK.

good question, I just expected it to work without thinking too much about it.
If that's the case, then maybe it doesn't make sense to try to support it.

@hn8 did you try to use TLS termination for HTTP on a load balancer forwading to TURN and it failed entirely or is there a possibility that it would work nevertheless?

**Comment by hn8** [comment]
@lukasIO I tried Envoy (as LB) and use two SNIs (one for WS and the other for TURN) so they can share TCP port 443. For WS/HTTP, TLS is terminated in LB. For TLS TURN, it uses TCP forwarding but TLS is terminated in LiveKit server. LB also forwards TCP-ICE, just not UDP TURN. I could not find filter options to do non-HTTP TLS-TCP forwarding, and obviously forwarding HTTP to TURN port didn't work as expected.

**Comment by davidzhao** [comment]
> @davidzhao @lukasIO Is it possible to terminate TLS and forward TCP for TURN since it is not HTTP? I could not make it work with Envoy before, only TLS forwarding (with SNI) works. Does it need special kind of protocol filter? Cloud load balancer only supports HTTP for TLS termination AFAIK.

Yeah, it's possible to terminate TLS first and then forward the underlying traffic over to LiveKit TURN. We've used AWS Network LB to terminate TLS there. I think it's possible to handle them on the same port/same cert as well: i.e. HTTP traffic flows to web port, while other TCP traffic goes to TURN/TCP)
